### DISCUSSION TITLE
How to uniformly fetch data for a page on both the client and server in Next.js App Router?

### DISCUSSION THREAD

**Discussion by lancejpollard** [vercel/next.js#69958] [Help]
### Summary

I have this in my `page.tsx` currently:

```tsx
import Page from '~/components/pages/languages/language/Page'
import { getAPI } from '~/data/base'

type Input = {
  params: { language: string }
}

export default async function View({ params }: Input) {
  const language = await getAPI(`/languages/${params.language}`)

  return <Page language={language} />
}
```

I also have a list view here:

```tsx
import Page from '~/components/pages/languages/Page'
import { getAPI } from '~/data/base'

export default async function View() {
  const languages = await getAPI(`/languages?size=1000`)

  return <Page languages={languages} />
}
```

This works fine as server-side rendered HTML, but every time I click on a link on the list page, to visit the item page, it takes about 1-1.5 seconds to change pages (and no sign of it changing during that time). I could try and add a `loading.tsx` to quickly show the loading state, but that has some issues with my layout component (beside the point for this post), so i wanted to see how I can make this feel more like a single-page app and _fetch the data client side_, so I can make a slightly faster refresh with some inline loaders.

Is there a pattern to wire up both client and server side data rendering? I have the server side rendering above like I mentioned, but how should I make it so it fetches from the API client-side as well (so it does both)? What are your recommendations (for App router style codebase)?

- I would like to just use the `<Link href />` Next.js component, instead of using some dynamic router system and manually binding to onClick of something.
- Or if that's not possible, maybe you could outline how this is generally accomplished?

### Additional information

```js
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.6.0: Mon Jul 29 21:14:30 PDT 2024; root:xnu-10063.141.2~1/RELEASE_ARM64_T6030
Binaries:
  Node: 20.10.0
  npm: 10.4.0
  Yarn: 1.22.19
  pnpm: 9.6.0
Relevant Packages:
  next: 14.0.4
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 4.9.5
Next.js Config:
  output: N/A
```

**Comment by Shriharsh-Deshmukh** [comment]
Hey **@lancejpollard** 

To make your page load faster and feel more like a single-page app while keeping the simplicity of using `<Link href />`, you can implement client-side data fetching alongside your existing server-side rendering. Here's a simplified approach:

1. **Switch to `useEffect` for client-side fetching**: When a link is clicked, instead of waiting for the server-side render, fetch the data on the client side for a faster perceived load.

2. **Preload API data**: You can preload the API data by using `React.useEffect` or `React Query` for smooth client-side data handling without extra routing.

3. **Add inline loaders**: Show loaders during the client-side fetching to improve user experience without changing your layout too much.

Here’s how to modify your component:

```tsx
import { useEffect, useState } from 'react'
import Page from '~/components/pages/languages/language/Page'
import { getAPI } from '~/data/base'

export default function View({ params }) {
  const [language, setLanguage] = useState(null)
  
  useEffect(() => {
    const fetchLanguage = async () => {
      const data = await getAPI(`/languages/${params.language}`)
      setLanguage(data)
    }
    
    fetchLanguage()
  }, [params.language])

  if (!language) {
    return <div>Loading...</div>
  }

  return <Page language={language} />
}
```

With this, you're fetching client-side while maintaining the server-side logic for SEO, and you'll get a faster page transition.

Let me know if this works for you!

**Reply by lancejpollard** [reply]
How do you use the `Link` in this case, I don't see how to use that still, and also do client side fetching on link click.
